kairos-prover/kairos-circuit-logic/src/account_trie.rs

```diff
@@ -0,0 +1,376 @@
+    #[allow(clippy::type_complexity)]
+    pub fn apply_batch(
+        &mut self,
+        transactions: impl Iterator<Item = KairosTransaction>,
+    ) -> Result<(Box<[L1Deposit]>, Box<[Signed<Withdraw>]>), TxnErr> {
+        let mut l1_deposits = Vec::new();
+        let mut l2_withdrawals = Vec::new();
+
+        for txn in transactions {
+            match txn {
+                KairosTransaction::Transfer(transfer) => {
+                    self.transfer(&transfer.public_key, &transfer.transaction, transfer.nonce)?;
+                }
+                KairosTransaction::Withdraw(withdraw) => {
+                    self.withdraw(&withdraw.public_key, &withdraw.transaction, withdraw.nonce)?;
+                    l2_withdrawals.push(withdraw);
+                }
+
+                KairosTransaction::Deposit(deposit) => {
+                    self.deposit(&deposit)?;
+                    l1_deposits.push(deposit);
+                }
+            }
+        }
+
+        Ok((
+            l1_deposits.into_boxed_slice(),
+            l2_withdrawals.into_boxed_slice(),
+        ))
+    }
+
+    /// Avoid calling this method with a transfer that will fail.
+    /// This method's prechecks may cause the Snapshot to bloat with unnecessary data if the transfer fails.
+    pub fn transfer(
+        &mut self,
+        sender: &PublicKey,
+        transfer: &Transfer,
+        nonce: u64,
+    ) -> Result<(), TxnErr> {
+        let [sender_hash, recipient_hash] =
+            hash_buffers([sender.as_slice(), transfer.recipient.as_slice()]);
+
+        let mut sender_account = self.txn.entry(&sender_hash)?;
+
+        let sender_account = sender_account.get_mut().ok_or_else(|| {
+            format!(
+                "sender does not have an account, sender: `{sender:?}`, sender_hash: `{sender_hash:?}`"
+            )
+        })?;
+
+        if sender_account.pubkey != *sender {
+            return Err(("hash collision detected on sender account").into());
+        }
+
+        sender_account.check_nonce(nonce)?;
+        sender_account.increment_nonce();
+
+        // Remove the amount from the sender's account
+        sender_account.balance = sender_account
+            .balance
+            .checked_sub(transfer.amount)
+            .ok_or("Sender balance underflow")?;
+
+        let recipient_account = self
+            .txn
+            .entry(&recipient_hash)?
+            .or_insert_with(|| Account::new(transfer.recipient.clone(), 0, 0));
+
+        if recipient_account.pubkey != transfer.recipient {
+            return Err(("hash collision detected on recipient account").into());
+        }
+
+        // Add the amount to the recipient's account
+        recipient_account.balance = recipient_account
+            .balance
+            .checked_add(transfer.amount)
+            .ok_or("recipient balance overflow")?;
+
+        Ok(())
+    }
+
+    pub fn deposit(&mut self, deposit: &L1Deposit) -> Result<(), TxnErr> {
```

Comment:
Why the choice to make this associated function take the `L1Deposit` struct directly but in `withdraw` and `transfer` you decompose the types. We could make it uniform and use the structs in `withdraw` and `transfer` too

Reply:
`L1Deposit` contains the recipient public key, and amount, it's only a payload.
The transfer and withdraw take a payload and the sender public key and a nonce.
The sender account will have the fund removed, and nonce checked/bumped. 
A deposit has no sender account, since it comes from the L1 so it's only a payload.

`fn transfer(&mut self, sender: &PublicKey, transfer: &Transfer, nonce: u64) -> Result<(), TxnErr>`
`fn withdraw(&mut self, withdrawer: &PublicKey, withdraw: &Withdraw, nonce: u64) -> Result<(), TxnErr>`